libopae++/include/opaec++/log.h

```diff
@@ -0,0 +1,215 @@
+ private:
+  std::stringstream sstream_;
+  int level_;
+  char fmt_[4];
+};
+
+/**
+ * @brief Named logger object used to format log messages with the process
+ * id, the logger name, and the level of the log message
+ */
+class logger {
+ public:
+  /**
+   * @brief log level enumeration that roughly maps to the log levels in
+   * libopae-c
+   */
+  enum class level : int
+  {
+      none        = -1,
+      error       = 0,
+      exception   = 10,
+      fatal       = 15,
+      warn        = 20,
+      info        = 25,
+      debug       = 40
+  };
+
+  /**
+   * @brief named logger constructor
+   *
+   * @param name The name of the logger to include in log messages
+   */
+  logger(const std::string & name);
+
+  /**
+   * @brief Generic log method that uses loglevel parameter
+   *
+   * @param l The level of the log message
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream log(level l, std::string str = "");
+
+  /**
+   * @brief Log method for debug messages
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream debug(std::string str = "");
+
+  /**
+   * @brief Log method for info messages
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream info(std::string str = "");
+
+  /**
+   * @brief Log method for warning messages
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream warn(std::string str = "");
+
+  /**
+   * @brief Log method for error messages
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream error(std::string str= "");
+
+  /**
+   * @brief Log method for exception messages
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream exception(std::string str = "");
+
+  /**
+   * @brief Log method for fatal error messages
+   *        Meant to be called when detecting a condition that will
+   *        terminate program execution
+   *
+   * @param str An optional string that may include more contextual
+   *            information
+   *
+   * @return A wrapped_stream object to begin the stream sequence
+   */
+  wrapped_stream fatal(std::string str = "");
+
+
+ private:
+  std::string name_;
+  int pid_;
```

Comment:
Should we use pid_t ?

Reply:
We could but that would require us to include `unistd.h` in this header. I try to keep the C++ API generic (without OS specific includes) if possible.